### ISSUE TITLE
make ExpressionsBasedModel.isInfeasible() public

### ISSUE THREAD

**Issue by drdv** [optimatika/ojAlgo#415]
This feature request is not related to a problem. I want to achieve something along the lines of the following example:

```
import org.ojalgo.netio.BasicLogger;
import org.ojalgo.optimisation.Expression;
import org.ojalgo.optimisation.ExpressionsBasedModel;
import org.ojalgo.optimisation.Optimisation;
import org.ojalgo.optimisation.Variable;

public class Example {
  public static Optimisation.State f() {
    ExpressionsBasedModel model = new ExpressionsBasedModel();

    Variable x = model.addVariable("x").weight(1.0);

    Expression c0 = model.addExpression("c0").upper(1);
    c0.set(x, 1);

    Expression c1 = model.addExpression("c1").lower(2);
    c1.set(x, 1);

    ExpressionsBasedModel model_simplified = model.simplify();

    if (model_simplified.isInfeasible()) {  // error: cannot be accessed from outside package
      return Optimisation.State.INFEASIBLE;
    } else {
      Optimisation.Result result = model_simplified.maximise();
      return result.getState();
    }
  }

  public static void main(final String[] args) {
    BasicLogger.debug(f());
  }
}
```
I have observed that `model.simplify()` returns an equivalent model when the model is either feasible or cannot be detected to be infeasible by the presolver. In this case  calling `model_simplified.maximise()` appears to be reasonable. However, if the presolver detects that the problem is infeasible, conflicting constraints are removed and the returned problem is no longer equivalent to the original one. In this case the result of `model_simplified.maximise()` is not useful for me. 

So I assume that the presolver is reliable and I want to return `INFEASIBLE`. There are two problems:

- The function `isInfeasible()` cannot be accessed as it is not public
- I am not familiar with `ojAlgo` and I am not sure that what I want to do is reasonable to begin with. 

Alternatives: Something we do currently is:
1. Call `model_simplified.maximise()`
2. If the result is infeasible exit
3. If the problem is not solved within a given time we assume that the problem admits a solution and call `model_simplified.maximise()`. The first solve, however, seems redundant and all this feels quite hacky.

The reason I want to `simplify()` is that using the presolver really makes a difference on the problem instances we are dealing with. 

Probably there is a standard/simpler/better/more reliable way to achieve this?

Thanks,
Dimitar

**Comment by apete** [comment]
If `model.simplify(`) returns a model that is not equivalent to the original, that's a bug.

Presolving is always done as part of `maximise()` or `minimise()`. There is no need to call `simplify()` unless you explicitly want the simplified model. Just calling `maximise()` on the original model I trust you get state `INFEASIBLE`.

I prefer to keep the `isInfeasible()` method private – it's for internal record keeping of what the presolver found. What you should check is the state.

**Comment by drdv** [comment]
Thanks for the clarification. Below I give an example (a slight modification of the one in my first post) that demonstrate that the simplified problem is not equivalent (or at least that is how I read it).

```
import org.ojalgo.netio.BasicLogger;
import org.ojalgo.optimisation.Expression;
import org.ojalgo.optimisation.ExpressionsBasedModel;
import org.ojalgo.optimisation.Optimisation;
import org.ojalgo.optimisation.Variable;

public class SimplificationProblemDemo {

  public static void main(final String[] args) {

    ExpressionsBasedModel model = new ExpressionsBasedModel();

    Variable x = model.addVariable("x").weight(1.0);

    Expression c0 = model.addExpression("c0").level(1);
    c0.set(x, 1);

    Expression c1 = model.addExpression("c1").level(2);
    c1.set(x, 1);

    ExpressionsBasedModel model_simplified = model.simplify();

    BasicLogger.debug("----------------------------------------------------");
    BasicLogger.debug("model:");
    BasicLogger.debug("----------------------------------------------------");
    BasicLogger.debug(model);
    Optimisation.Result res1 = model.minimise();
    BasicLogger.debug(res1);  // INFEASIBLE (as expected)
    BasicLogger.debug(model);

    BasicLogger.debug("----------------------------------------------------");
    BasicLogger.debug("model.simplify():");
    BasicLogger.debug("----------------------------------------------------");
    BasicLogger.debug(model_simplified);
    Optimisation.Result res2 = model_simplified.minimise();
    BasicLogger.debug(res2);  // DISTINCT!
    BasicLogger.debug(model_simplified);
  }

}
```
The output:

```
----------------------------------------------------
model:
----------------------------------------------------
############################################
1 <= x: 1 (1) <= 1
1 <= c0: 1.0 <= 1
2 <= c1: 1.0 <= 2
############################################

INFEASIBLE 1.0 @ { 1 }
############################################
1 <= x: 1 (1) <= 1
1 <= c0: 1.0 <= 1
2 <= c1: 1.0 <= 2
############################################

----------------------------------------------------
model.simplify():
----------------------------------------------------
############################################
1 <= x: 1 (1) <= 1
############################################

DISTINCT 1.0 @ { 1 }
############################################
1 <= x: 1 (1) <= 1
############################################
```

**Comment by drdv** [comment]
Of course, the above example is only for demonstration purposes. When dealing with "useful" problems I have observed that manually calling `simplify()` has a huge impact on the time it takes to solve the problem. But this shouldn't be the case (based on what you say) - or maybe I am missing something?

**Comment by apete** [comment]
If you compare

`model.simplify().maximise();`

with simply

`model.maximise();`

is there a significant time difference?

**Comment by drdv** [comment]
The problem (BLP): 1657 boolean variables, 164 constraints

1. With ojAlgo v. 49.0.3, and a 30 sec. time limit:
- `model.maximise()` takes 30 sec., the status is `FAILED`
- `model.simplify().maximise()` takes 1s, the status is `OPTIMAL`, the value of the objective function is the same as the one from CBC, GLPK

2. Now I decided to test with the latest version (51.0.0) and to my surprise the behaviour is different: both problems (with and w/o `simplify()`) take 30s and return a flag `FEASIBLE`. The objective value is very close to the reference in both cases (~0.03% off).

I suppose that there must have been a change in the presolver between version 49.0.3 and 51.0.0.

**Comment by apete** [comment]
Several things have changed since v49.0.3. In particular v49.1.0 and v49.2.0 seems to have relevant changes:
https://github.com/optimatika/ojAlgo/blob/develop/CHANGELOG.md

When an integer model returns state `FEASIBLE` it should mean that an integer solution is found but not yet "proved" that there is no better. I've seen other cases where the optimal integer solution actually has been found, but the algorithm still needs to continue for a very long time to.

You can set options to get some runtime info:
`model.options.progress(IntegerSolver.class);`
`model.options.debug(IntegerSolver.class);` (will most likely be far too much info with a large model like this)

Would be interesting to know why v49.0.3 and `model.simplify().maximise()` finished so much faster. Perhaps you can see something with that runtime info.

With v51.0.0 it is possible to write a model to file `model.writeTo(File)`. Can you give me a model file?

There is a branch you can build from to get a version with the infeasible simplification bug fixed: https://github.com/optimatika/ojAlgo/tree/InfeasibleSimplification

**Comment by drdv** [comment]
Thanks for the bugfix, I will test with it.

**Comment by apete** [comment]
Not sure what to say or do. As far as I can see the MIP solver does eventually find the solution. It's just very slow.

The fact that `model.simplify().maximise()` got the correct solution in just 1s with v49.0.3 is surprising. Especially since `model.maximise()` couldn't find any integer solution at all within 30s. Running in a debugger, with progress logging turned on, can you give me more info about how this happened?

What I see with v51.0.0 is that it takes many seconds before the first integer solution is found. Shortly after that a sequence of incrementally improved solutions are found, but the objective function value only changes in the 4th or 5th digit. In terms of the objective function value the first and final integer solutions found are very similar.

BTW what is the expected objective function value when maximising the model you supplied? The best ojAlgo can find is: 139.4070725458

**Comment by drdv** [comment]
Actually I just realised that (probably) it is not possible to import the `.ebm` file I sent with `v49.0.3`. So since I am sure that you would prefer to debug it yourself I am sending an `.mps` file with the same problem ([problem_to_send.mps.txt](https://github.com/optimatika/ojAlgo/files/8147015/problem_to_send.mps.txt)). 

CBC solves the problem in ~350ms and the objective value is (exactly) the same as yours (that is why I thought that 1s is quite reasonable).

Regarding you bugfix: seems to work just fine (I don't observe anymore the behaviour I described in my first post).

BTW, could you point me to (detailed) documentation of all possible exit statuses (for example, what is the difference between `APPROXIMATE` and `FEASIBLE`).

**Comment by apete** [comment]
OK, now I understand. There was a bug in v49.0.3, another problem with the `simplify()` method. It also removed the integer requirements (which is what it's supposed to do when used internally by the `IntegerSolver`).

With v49.0.3

`model.maximise()`; Not able to find any integer solution within the given time limit, and therefore returns state `FAILED`.

`model.simplify().maximise()`; Does not use the `IntegerSolver`, and instead uses the `LinearSolver`. The returned solution is NOT an integer solution, but a solution to the relaxed linear problem. The objective function value is 139.4144891974483. You say that this is the same as what CBC/GLPK returns. Is it? Is it really exactly the same, and is that solution an integer solution?

With v51.0.0

There is no difference between calling `model.maximise()` and `model.simplify().maximise()`. The `IntegerSolver` is used correctly. It takes something like 30s to get the first integer solution, and 30min to find/prove the final optimal integer solution. That solution is 139.4070725458.

As you can see the relaxed linear solution and the integer solution only differ in the 6th digit (rounding to 5 digits they're equal). This is part of the difficulty in solving the MIP.

If you use v51.0.0 and don't call `simplify()` explicitly, ojAlgo functions correctly, but takes very long to solve this problem.

This is a fairly large binary model. Just the fact that it takes a very long time to solve is not a bug. For sure ojAlgo can be improved, but that's time consuming work that needs to be funded for it to happen.

Regarding documentation of the `Optimisation.State` enum: It has javadoc and you have the source code. There is no other documentation.

**Comment by drdv** [comment]
Interesting. This explains what I observe. 

A few comments:
- You are right, with `v49.0.3` the  solution is not integer for `model.simplify().maximise()` - I have overlooked that (when exporting the problem I have an assumption of integer variables and that is why CBC ...)

- Regarding `v51.0.0` (just a clarification): `model.maximise()` and `model.simplify().maximise()` result in the same behaviour on this very problem, in general for infeasible problems they don't (but this is solved in `v51.0.1`).

- Documentation of the `Optimisation.State`: I was looking at the wrong place (things are clear now).

This issue can be closed.

A final question: where can I find details about the algorithm used in `ojAlgo` and do you have a list of things you believe can improve it. 

Thanks for the support.

**Comment by apete** [comment]
There is no documentation detailing the various optimisation algorithms used.

If you're interested in trying to improve things I'll be happy to guide you getting started, and I do have a list of things that could be worth looking in to.

I'll close this issue now. Perhaps start a thread in the discussions area: https://github.com/optimatika/ojAlgo/discussions
